Support for "dashed" polylines

jasonbeverage opened the pull request:
This PR adds support for screen space "dashed" polylines.  This attempts to get something similar to OpenGL line stippling to work with Cesium's neat wide polyline algorithm.  I tried various techniques using the texture coordinates of the line but none ended up looking as nice as this one.

There is a new PolylineDashMaterialProperty that controls the PolylineDash material.

I've modified the PolylineMaterialAppearanceVS.glsl shader to compute a varying, v_angle, that is fed to the fragment shader.  This angle is the screen space angle between the line and the x axis quantized to 45 degree increments.  I've quantized rather than using the actual value to try to avoid artifacts as the polyline expands.

The fragment shader takes the incoming pixel coordinate, rotates it by the incoming angle and then does a simple stippling along the x-axis on the rotated pixel coordinate.  The fragment location is rotated to create a striped pattern in the direction of the line so it actually appears like it's being stippled along the line.

I've updated the Polyline and Polyline CZML sandcastle examples to have a dashed line so you can see it in action.

![dash](https://cloud.githubusercontent.com/assets/326626/24376697/d4ae324e-130a-11e7-9f10-1262a2366d04.jpg)

hpinkos commented:
This is awesome @jasonbeverage, thanks!  This has been a frequently requested feature and will make the Cesium community very happy :smile:

pjcozzi commented:
Fixes #2584

pjcozzi commented:
Thanks @jasonbeverage.  This is a pleasant surprise and will be a very popular feature!

I would like to review this before the final merge.  @bagnell should also take a look at this first since he wrote the original polyline shader, but he's busy for probably all of this week.  In the meantime, if anyone has cycles to review or even just test this, please jump in!

@jasonbeverage I'm not sure if you've already read it, but you might like this paper: [Shader-Based Antialiased, Dashed, Stroked Polylines](http://jcgt.org/published/0002/02/08/)

mramato commented:
Awesome work @jasonbeverage. You keep cranking out some great stuff. I'm not qualified to review the shaders here, but the Entity and CZML pieces look spot on (will look closer as soon as I have time if no one else beats me to it).

One of the things me and @bagnell have talked about for stippled lines is the ability to specify a custom pattern (perhaps through a string like `-- - ---`).  Would that be an easy extension to this implementation?  (I'm just asking out of curiosity).  I know @bagnell also had some ideas (such as using images as part of line styles but that's also outside the scope of this PR).

emackey commented:
This is crazy awesome.

@mramato You probably know shaders can't directly take a string, but we could pass a float parameter to indicate a percentage of dash vs. gap, or I could even imagine passing a one-dimensional alpha map that supplies a dashing pattern to use.  That would be for a future version of course, this seems cool enough as-is that it should probably go in without a lot of feature creep.  Also I was thinking it could be useful to supply a second color for use in the gaps, instead of just discarding them (for example, use the same color with less alpha, or alternate the brightness, etc).  But again, that's feature creep, and this is plenty good enough already.

Dashed lines (as they are now) will be very useful in combination with the depthFail feature going on in #5160.  Underground lines are crying out for dashes.

jasonbeverage commented:
Don't merge this just yet, I'm going to try to push a couple more things and see what you think.

emackey commented:
@jasonbeverage great.  If you're making changes can you switch `Color.BLUE` to `Color.CYAN` to make the line more visible?  Likewise in the CZML crank the green channel to 255.  Thanks!

jasonbeverage commented:
Sure thing, just pushed that change.

emackey commented:
I got asked by @GatorScott to test with a more curved path, and I had an inclined GEO handy.  [Here's a gist](https://gist.github.com/anonymous/44bf911c2d312aee84446805a33d08c0).  It works pretty well, even for this tough case.

jasonbeverage commented:
I just pushed a few changes.

There is a new dashPattern property for the PolylineDashMaterial that defines a 16 bit stipple pattern.  

I changed some of the defaults so that dashLength is 16 pixels so it lines up per pixel (approximately) with the dash pattern.  The default dash pattern is 255 (0000 0000 1111 1111).

I've also added a new Polyline Dash sandcastle example.

Also, just wanted to say thanks to @timoore for doing the initial work on this effort, including the dashPattern logic.  Our original approach just used the distance along the line to drop pixels, but that caused too many artifacts with your wide polyine algorithm (similar to what you'd see with the StipeMaterial with the direction set to vertical).  I came up with the screen space fragment shader based stippling just recently and the artifacts aren't nearly as bad.

![polyline dash](https://cloud.githubusercontent.com/assets/326626/24426105/78da5478-13d4-11e7-8438-170625b3762b.jpg)

emackey commented on Source/DataSources/PolylineDashMaterialProperty.js:
```diff
@@ -0,0 +1,142 @@
+        }
+        result.color = Property.getValueOrClonedDefault(this._color, time, defaultColor, result.color);
+        result.dashLength = Property.getValueOrDefault(this._dashLength, time, defaultDashLength, result.dashLength);
+        result.dashPattern = Property.getValueOrDefault(this._dashPattern, time, defaultDashPattern, result._dashPattern);
```
I don't think `result._dashPattern` gets an underscore here.

emackey commented on Source/DataSources/PolylineDashMaterialProperty.js:
```diff
@@ -0,0 +1,142 @@
+     * @param {Object} [options] Object with the following properties:
+     * @param {Property} [options.color=Color.WHITE] A Property specifying the {@link Color} of the line.
+     * @param {Property} [options.dashLength=16.0] A numeric Property specifying the length of the dash pattern in pixel.s
+     * @param {Property} [options.dashPattern=255.0.0] A numeric Property specifying a 16 bit pattern for the dash
```
Typo

emackey commented:
I wonder if CZML (and maybe Entity) can take a 16-character string instead of numeric `dashPattern`.  For example:

```
var dashPattern = '        --------';
```

Could this be converted to 255 by testing each char position and setting a bit?  This would make the pattern more user-friendly.

emackey commented:
Here's some code to turn an arbitrary-length string of spaces and dashes into a closest-approximation 16-bit field.  This just iterates the 16 bits, picks the closest char based on length of input string, and figures out if it will set that bit or not.

```
function dashStringToBits(dashPatternString) {
        var dashPatternLength = dashPatternString.length;
        var dashPatternNumber = 0;
        var currentBitValue = 65536;
        for (var i = 0; i < 16; ++i) {
            currentBitValue = currentBitValue >> 1;
            var charPos = Math.floor((i / 16) * dashPatternLength);
            var char = dashPatternString.charAt(charPos);
            if (char !== ' ') {
                dashPatternNumber |= currentBitValue;
            }
        }
        return dashPatternNumber;
    }
```

```
dashStringToBits(' -')
255
dashStringToBits('- ')
65280
dashStringToBits('    ----')
255
dashStringToBits('        --------')
255
dashStringToBits('  -  ---')
3135
```

jasonbeverage commented:
Just pushed a fix for my fat finger typos :)

emackey commented:
> Would PolylineDashMaterialProperty.getValue just need to call that 

Off the top of my head I'm not sure where it goes.  I'd have to research a bit.  Curious to hear @mramato's thoughts before we go too far on it.

jasonbeverage commented:
I've just been mashing 0 and 1 into the Windows programmer calculator and copying the resulting decimal value :)  

I also just tried your gist and it looks awesome!

pjcozzi commented:
> Also, just wanted to say thanks to @timoore for doing the initial work on this effort, including the dashPattern logic...

Thanks again to both of you!

mramato commented:
While I was an original proponent of the string technique, it's actually not needed.  There are two easy ways just use binary literals in JavaScript.

1. In "old" JS, you can use `parseInt`, for example: `parseInt('1010101010101010', 2)`.
2. ES6 has binary literals with the `0b` as a prefix.  For example: `0b0000111100001111`.

We should probably update the example to use 1, since it will work everywhere, but the doc example might want to reference both.

I think these are easy enough to grok that the string approach is obsolete.

For CZML, you would need the number in the data instead of a string, but the writer could easily handle this rather than forcing the client to do it (plus sending a number down is way more efficient than putting a string in CZML).  What do you guys think?

pjcozzi commented:
+1 for @mramato's ideas.  This is also how we would represent it in 3D Tiles.

jasonbeverage commented:
I like it.  I replaced the explicit dashPattern in the example with parseInt, much easier to read.

emackey commented:
What do you think about adding a `gapColor` instead of `discard`?

```
float maskTest = floor(dashPattern / pow(2.0, maskIndex));
vec4 fragColor = (mod(maskTest, 2.0) < 1.0) ? gapColor : color;
if (fragColor.a < 0.005) {   // matches 0/255 and 1/255
    discard;
}

material.emission = fragColor.rgb;
material.alpha = fragColor.a;
return material;
```

pjcozzi commented:
+1 for `gapColor`!

jasonbeverage commented:
I just pushed the new gapColor property and updated the Polyline Dash example to demonstrate it.

emackey commented on Source/DataSources/PolylineDashMaterialProperty.js:
```diff
@@ -28,6 +29,7 @@ define([
      *
      * @param {Object} [options] Object with the following properties:
      * @param {Property} [options.color=Color.WHITE] A Property specifying the {@link Color} of the line.
+     * @param {Property} [options.gapColor=Color.BLACK.withAlpha(0.0)] A Property specifying the {@link Color} of the gaps in the line.
```
Use `Cesium.Color.TRANSPARENT`

emackey commented on Source/Shaders/Appearances/PolylineMaterialAppearanceVS.glsl:
```diff
@@ -10,20 +10,45 @@ attribute float batchId;
+    else {
+        dir = normalize(nextClip.xy - pClip.xy);
+    }
+    v_angle = atan(dir.x, dir.y) - atan(1.0, 0.0);
```
So I have to ask:  Are there any performance concerns here @pjcozzi @bagnell?  There's a bunch of new math, here and in `PolylineVS.glsl`, that appears to run for all polylines, not just dashed ones.  It's all vertex shader, not fragment shader math, so maybe that's no problem?  I may have a test file with a lot of polylines, I'll see if I can find that.

emackey replied:
OK, I found a 168 meg CZML file full of polylines, that's quite a stress test bringing Cesium down to about 5fps on my Quadro K2000M.  I tried this with the performance counter turned on, and it appears to hover around 170ms per frame with or without this branch.  So, for my card at least, there's no performance concern.

emackey commented:
It's sooo tempting to merge this for 1.32.  Someone remind me why we shouldn't.

jasonbeverage commented:
You could change the getPolylineWindowCoordinates function in PolylineCommon to have an out angle parameter to have a performance boost and avoid a little bit of duplicate work.  I won't pretend to understand exactly what is going on in there, but you can do something like this:

```
vec4 getPolylineWindowCoordinates(vec4 position, vec4 previous, vec4 next, float expandDirection, float width, bool usePrevious, out float angle) {
    vec4 endPointWC, p0, p1;
    bool culledByNearPlane, clipped;

    vec4 positionEC = czm_modelViewRelativeToEye * position;
    vec4 prevEC = czm_modelViewRelativeToEye * previous;
    vec4 nextEC = czm_modelViewRelativeToEye * next;

    vec4 positionWC = czm_eyeToWindowCoordinates(positionEC);
    vec4 pWC = czm_eyeToWindowCoordinates(prevEC);
    vec4 nWC = czm_eyeToWindowCoordinates(nextEC);

    clipLineSegmentToNearPlane(prevEC.xyz, positionEC.xyz, p0, clipped, culledByNearPlane);
    clipLineSegmentToNearPlane(nextEC.xyz, positionEC.xyz, p1, clipped, culledByNearPlane);
    clipLineSegmentToNearPlane(positionEC.xyz, usePrevious ? prevEC.xyz : nextEC.xyz, endPointWC, clipped, culledByNearPlane);

    vec2 prevWC = normalize(p0.xy - endPointWC.xy);
    vec2 nextWC = normalize(p1.xy - endPointWC.xy);

    // Determine the relative screen space direction of the line.
    vec2 lineDir;
    if (usePrevious) {
        lineDir = normalize(positionWC.xy - pWC.xy);
    }
    else {
        lineDir = normalize(nWC.xy - positionWC.xy);
    }
    angle = atan(lineDir.x, lineDir.y) - atan(1.0, 0.0);

    // Quantize the angle so it doesn't change rapidly between segments.
    angle = floor(angle / czm_piOverFour + 0.5) * czm_piOverFour;

    if (culledByNearPlane)
    {
        return vec4(0.0, 0.0, 0.0, 1.0);
    }

    float expandWidth = width * 0.5;
    vec2 direction;

    if (czm_equalsEpsilon(previous.xyz - position.xyz, vec3(0.0), czm_epsilon1) || czm_equalsEpsilon(prevWC, -nextWC, czm_epsilon1))
    {
        direction = vec2(-nextWC.y, nextWC.x);
    }
    else if (czm_equalsEpsilon(next.xyz - position.xyz, vec3(0.0), czm_epsilon1) || clipped)
    {
        direction = vec2(prevWC.y, -prevWC.x);
    }
    else
    {
        vec2 normal = vec2(-nextWC.y, nextWC.x);
        direction = normalize((nextWC + prevWC) * 0.5);
        if (dot(direction, normal) < 0.0)
        {
            direction = -direction;
        }

        // The sine of the angle between the two vectors is given by the formula
        //         |a x b| = |a||b|sin(theta)
        // which is
        //     float sinAngle = length(cross(vec3(direction, 0.0), vec3(nextWC, 0.0)));
        // Because the z components of both vectors are zero, the x and y coordinate will be zero.
        // Therefore, the sine of the angle is just the z component of the cross product.
        float sinAngle = abs(direction.x * nextWC.y - direction.y * nextWC.x);
        expandWidth = clamp(expandWidth / sinAngle, 0.0, width * 2.0);
    }

    vec2 offset = direction * expandDirection * expandWidth * czm_resolutionScale;
    return vec4(endPointWC.xy + offset, -endPointWC.z, 1.0);
}
```

I don't think you can use the "clipped" window coordinates in this situation.  Every time I've tried I end up with artifacts.

jasonbeverage commented:
I went ahead and tried to consolidate the code in getPolylineWindowCoordinates.  This avoids a couple of duplicate operations and also cleans up the polyline shaders so they don't have duplicated code.  Give it a look and see what you think.

pjcozzi commented:
> It's sooo tempting to merge this for 1.32. Someone remind me why we shouldn't.

I'd like @bagnell and me to look at this first.  We are on a few other projects right now so I don't know that we can realistically do this by Monday.

> OK, I found a 168 meg CZML file full of polylines, that's quite a stress test bringing Cesium down to about 5fps on my Quadro K2000M. I tried this with the performance counter turned on, and it appears to hover around 170ms per frame with or without this branch. So, for my card at least, there's no performance concern.

The bottleneck will depend on lots of factors like screen size, number of vertices, view direction, etc.  We'd like this to be as efficient as possible, especially if we build 3D Tiles vector tiles on this.

We usually just `#ifdef` parts of the shader out when it is not being used.  For example, see
* [BillboardCollectionVS.glsl](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Shaders/BillboardCollectionVS.glsl)
* [BillboardCollection.js](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/BillboardCollection.js)

jasonbeverage commented:
No rush to get it merged, I just wanted it out of my fork so you all could
see it :)

I've also got another PR that is even cooler but need a bit to clean it up
;)

On Wed, Mar 29, 2017, 4:54 PM Patrick Cozzi <notifications@github.com>
wrote:

> It's sooo tempting to merge this for 1.32. Someone remind me why we
> shouldn't.
>
> I'd like @bagnell <https://github.com/bagnell> and me to look at this
> first. We are on a few other projects right now so I don't know that we can
> realistically do this by Monday.
>
> OK, I found a 168 meg CZML file full of polylines, that's quite a stress
> test bringing Cesium down to about 5fps on my Quadro K2000M. I tried this
> with the performance counter turned on, and it appears to hover around
> 170ms per frame with or without this branch. So, for my card at least,
> there's no performance concern.
>
> The bottleneck will depend on lots of factors like screen size, number of
> vertices, view direction, etc. We'd like this to be as efficient as
> possible, especially if we build 3D Tiles vector tiles on this.
>
> We usually just #ifdef parts of the shader out when it is not being used.
> For example, see
>
>    - BillboardCollectionVS.glsl
>    <https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Shaders/BillboardCollectionVS.glsl>
>    - BillboardCollection.js
>    <https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/BillboardCollection.js>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AnalyticalGraphicsInc/cesium/pull/5159#issuecomment-290222586>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAT74tWPXetJWfhk6Gb1aM-uo9oL6MMGks5rqsT1gaJpZM4MqyVr>
> .
>

mramato commented:
> I've also got another PR that is even cooler but need a bit to clean it up

Now you're just teasing us 😄

bagnell commented on Source/Shaders/PolylineCommon.glsl:
```diff
@@ -32,30 +32,48 @@ void clipLineSegmentToNearPlane(
-    
+
+    // Compute the window coordinates of the points.
+    vec4 positionWindow = czm_eyeToWindowCoordinates(positionEC);
```
Can you `#ifdef` this out so polylines that don't use the angle don't compute it? For example, surround this code in `#ifdef POLYLINE_DASH` and in `PolylineDashMaterial.glsl` add `#define POLYLINE_DASH`.

bagnell commented:
I only have that one comment. Looks good! Thanks @jasonbeverage!

I believe @lilleyse also wants to review this.

bagnell commented:
@jasonbeverage I spent some time playing with your branch and the PolylineDashMaterial shader today. I created a branch ([dashedlines-2](https://github.com/AnalyticalGraphicsInc/cesium/tree/dashedlines-2) that has your changes plus some of my own. It adds a length in meters from the first point on the line as a new vertex attribute and I tried to compute `dashPosition` from that. This way you won't see the abrupt changes from `v_angle` or much of a change in the dash pattern when panning from using `gl_FragCoord`. It looks pretty good when panning or rotating as long as the camera is mostly looking top-down, but as soon as you start rotating, certain camera views will cause the dashes to shrink or stretch. Any ideas you have about this would be greatly appreciated!

I still want to see this PR merged into master. Computing the dashes based on line length can sit in a branch until someone has time to work on it.

pjcozzi commented on Source/DataSources/PolylineDashMaterialProperty.js:
```diff
@@ -0,0 +1,156 @@
+     * @returns {Boolean} <code>true</code> if left and right are equal, <code>false</code> otherwise.
+     */
+    PolylineDashMaterialProperty.prototype.equals = function(other) {
+        return this === other || //
```
Is the trailing `//` here and below required?

jasonbeverage replied:
Hey @pjcozzi it's not necessary and I've just pushed a change removing it.

pjcozzi commented on Source/Shaders/PolylineCommon.glsl:
```diff
@@ -32,30 +32,48 @@ void clipLineSegmentToNearPlane(
+    else {
+        lineDir = normalize(nextWindow.xy - positionWindow.xy);
+    }
+    angle = atan(lineDir.x, lineDir.y) - atan(1.0, 0.0);
```
`atan` can be a performance killer.
* Can you precompute `atan(1.0, 0.0)`?  I would not trust all GLSL compilers to optimize that.
* What level of precision is required for this?  Would it be possible to replace the other `atan` call with a taylor series expansion of a few terms?  Seems like yes since it is being quantized.

jasonbeverage replied:
I pushed a change precomputing the atan(1.0,0.0).  We could use a taylor series or a lookup table for the other atan, but I'd save that for later unless performance is a killer.

pjcozzi replied:
Last time I tested - which was awhile ago - atan is pretty much the worse thing we can do in a shader.  @bagnell said there was already inverse trig in the polyline shader, but I don't see it at quick glance.

If you're not up for it now, we'll try to make this change soon.

jasonbeverage commented:
Hey @bagnell , we tried almost the exact same thing initially, and I think it would work fine if you were just drawing regular old lines.  But I couldn't figure out a way to get it to play nice with the wide polyline aglo.  I think the problem is that the length in meters actually changes along the line when you move the verts around, so you end up with funky artifacts depending on how the line rotates.  Your algo also does something along the near plane where it looks like vert is pushed really far towards the near plane so you end up stretching the distance in meters artificially resulting in really wide dashes.

I tried and tried but couldn't come up with something that worked so I ended up scrapping it in favor of this purely fragment based stippling.

jasonbeverage commented:
@bagnell I tried wrapping the block in polylinecommon with the ifdef, but it looks like just adding #define POLYLINE_DASH to the material didn't work.  It behaves like it's not actually set.  Is there somewhere else the define would need to go?

denverpierce commented:
@mramato for CZML, I'd think it'd be easier and more intuitive to write a bit mask instead of a 16bit number, especially for users more familiar with the data side of things.

IE:
```
{
dashLength:16,
dashPatthern: '0000000011111111'
}
```

is more human read/writeable than:

`dashPattern: 255`

mramato commented:
@bagnell and @pjcozzi do you have any additional comments here?  I want to take another pass on the Entity updates (just because I haven't looked at this in a few weeks) but I don't expect any surprises.

Barring any issues, I would like to get this merged ASAP so that it has time to stew before 1.32 on May 1st.

Thanks.

pjcozzi commented:
@mramato review and merge when ready.

bagnell commented:
> @bagnell I tried wrapping the block in polylinecommon with the ifdef, but it looks like just adding #define POLYLINE_DASH to the material didn't work. It behaves like it's not actually set. Is there somewhere else the define would need to go?

Sorry, it needs to be in the vertex shader. Can you modify the `PolylineMaterialAppearance.vertexShaderSource` getter to search the material `shaderSource` for `varying float v_angle` and then append the `#define`?

mramato commented:
> @mramato for CZML, I'd think it'd be easier and more intuitive to write a bit mask instead of a 16bit number, especially for users more familiar with the data side of things.

We can probably add this in as a CZML-only feature in a small PR after this gets merged.  We'll also need to update the spec and writer too. (Of course we need that for the new dash stuff as well).  CC @shunter

pjcozzi commented:
@jasonbeverage we'd like to merge this in the next few days so it has some time in master before it ships in Cesium 1.33 on May 1.  Can you make the change @bagnell suggested, https://github.com/AnalyticalGraphicsInc/cesium/pull/5159#issuecomment-294560604, or do you want us to?

Does anyone else want to test or review this?

jasonbeverage commented:
If you can do it that would help me out, I'm knee deep in other stuff right
now and you'll be faster at it anyway :)

On Thu, Apr 20, 2017 at 10:55 AM Patrick Cozzi <notifications@github.com>
wrote:

> @jasonbeverage <https://github.com/jasonbeverage> we'd like to merge this
> in the next few days so it has some time in master before it ships in
> Cesium 1.33 on May 1. Can you make the change @bagnell
> <https://github.com/bagnell> suggested, #5159 (comment)
> <https://github.com/AnalyticalGraphicsInc/cesium/pull/5159#issuecomment-294560604>,
> or do you want us to?
>
> Does anyone else want to test or review this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AnalyticalGraphicsInc/cesium/pull/5159#issuecomment-295766567>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAT74gT0RL6pFeyMJwLjfO6K7fdZy8K0ks5rx3HSgaJpZM4MqyVr>
> .
>

bagnell commented:
@mramato Can you merge when ready?

denverpierce commented:
I threw some fairly complex linework at it and it performed admirably.

mramato commented:
Thanks again @jasonbeverage!

jasonbeverage commented:
Thanks for merging this!  Glad to be able to contribute.

pjcozzi commented:
Fantastic!  This will quickly become a popular feature.  Thanks again @jasonbeverage!

initialed85 commented:
@jasonbeverage I had a thirst for dashed lines that you have now quenched, thankyou.

Tezraine commented:
I know I'm a bit late to the party, but I just wanted to mention that with ECMA6, the 0b format for numbers is supported. Also 0b1111111100000000 (255) if reversed and overlapped is effectively a bitwise OR that leaves you with all 1s. 0b1111000000001111 (61455) or 0b0000111111110000 (4080) is the same pattern, but is the same forwards and backwards.

mickae1 commented:
Hello, now there is a problem when the camera is moving, the dash is also moving, which is not nice, I've a made a first pull request to fix that : 

https://github.com/CesiumGS/cesium/pull/12343

After this pull request validated, we can use the distance from the first point to calculate better the pattern :)

it's working on my computer
